lv_draw/lv_draw_label.c

```diff
@@ -54,148 +52,214 @@ static uint8_t hex_char_to_num(char hex);
-                            r = (hex_char_to_num(buf[0]) << 4) + hex_char_to_num(buf[1]);
-                            g = (hex_char_to_num(buf[2]) << 4) + hex_char_to_num(buf[3]);
-                            b = (hex_char_to_num(buf[4]) << 4) + hex_char_to_num(buf[5]);
-                            recolor = LV_COLOR_MAKE(r, g, b);
-                        } else {
-                            recolor.full = style->text.color.full;
-                        }
-                        cmd_state = CMD_STATE_IN; /*After the parameter the text is in the command*/
-                    }
-                    continue;
-                }
-            }
-
-            lv_color_t color = style->text.color;
-
-            if(cmd_state == CMD_STATE_IN) color = recolor;
-
-            letter_fp(&pos, mask, font, letter, color, opa);
-            letter_w = lv_font_get_width(font, letter);
-
-            pos.x += letter_w + style->text.letter_space;
-        }
-        /*Go to next line*/
-        line_start = line_end;
-        line_end += lv_txt_get_next_line(&txt[line_start], font, style->text.letter_space, w, flag);
-
-        pos.x = coords->x1;
-        /*Align to middle*/
-        if(flag & LV_TXT_FLAG_CENTER) {
-            line_width = lv_txt_get_width(&txt[line_start], line_end - line_start,
-                                          font, style->text.letter_space, flag);
-
-            pos.x += (lv_area_get_width(coords) - line_width) / 2;
-
-        }
-        /*Align to the right*/
-        else if(flag & LV_TXT_FLAG_RIGHT) {
-            line_width = lv_txt_get_width(&txt[line_start], line_end - line_start,
-                                          font, style->text.letter_space, flag);
-            pos.x += lv_area_get_width(coords) - line_width;
-        }
-
-        /*Go the next line position*/
-        pos.y += lv_font_get_height(font);
-        pos.y += style->text.line_space;
-    }
+	/*Write out all lines*/
+	while (txt[line_start] != '\0') {
+		if (offset != NULL) {
+			pos.x += x_ofs;
+		}
+		/*Write all letter of a line*/
+		cmd_state = CMD_STATE_WAIT;
+		i = line_start;
+		uint32_t letter;
+		// my code
+		uint32_t previous_letter;
+		uint32_t current_letter;
+		reversed_index = 0;
+		uint8_t noWidth = 0;
+		// end my code
+		while (i < line_end) {
+			// my code
+			//if it's necessary make the txt buffer reversed
+			if (!reversed_index) {
+				reversed_buffer = lv_get_reversed_buffer((uint8_t*) txt, i,
+						line_end);
+				previous_letter = 0;
+			}
+			//if the returned buffer is the same as txt buffer means no persian/arabic letter was found
+			if ((const char*)reversed_buffer == txt) {
+				letter = lv_txt_encoded_next(txt, &i);
+				reversed_index = i;
+			} else {
+				letter = lv_txt_encoded_next((const char*) reversed_buffer,
+						&reversed_index);
+				current_letter = letter;
+				// update i
+				i = line_start + reversed_index;
+				if ((letter > 0x600) && (letter < 0x6ff)) {
```

Comment:
What is the range 0x600 - 0x6ff? Is that related to Persian letters encoding? Is there a way to generalize it to other RTL languages?

Reply:
This has been explained in previous comments

Reply:
@pioneer2637,
My opinion is, that instead of specifying unicode ranges hard-coded in different places in your code, you could call a function (something like `bool lv_is_rtl_letter(uint32_t letter)`) and make the checks there.  
This would make porting to other languages easier. All we would have to do to support another language is to change `lv_is_rtl_letter` and add another range there.

Reply:
@amirgon It really sounds flexible. A default Persian `lv_is_rtl_letter` can be added to show an example.